### DISCUSSION TITLE
Refactor docker binarySource to use bash file instead of concatenated commands

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#21685] [Suggest an Idea]
We currently do like `docker run command1 && command2 && command3` but it's quite hard to escape it all and make sense. Perhaps writing these commands one per line to a bash file and executing that inside the container would be better?

**Comment by rarkins** [comment]
@viceice @zharinov what do you think?

**Comment by zharinov** [comment]
As the first step, maybe we can log script code instead of long escaped strings. It will help to debug artifact updates and we can reuse it for file emitting.

**Comment by rarkins** [comment]
Just a logging enhancement?

**Comment by zharinov** [comment]
In the second step, we'll redirect this generated string to the file and execute it

**Comment by viceice** [comment]
what about my idea i suggested some time ago?

* start container on demand with `docker -t -n renovate_node .... renovate/node cat -` 
* use `docker exec -e HHH=xxx ... renovate_node npm install -g yarn` to run commands one-by-one
* stop all renovate sidecar container after repo finished

*pro*
* can reuse container for multiple branches
* can use in conatiner cache for multiple branches

**Comment by rarkins** [comment]
I like the *pro* points and would like to try this. I guess a *con* point is:
- All sidecar containers will remain running/using resources until the repo is finished

We could maybe add the option to stop them after each exec() finishes if anyone has a problem with this.

**Comment by viceice** [comment]
sure, we can stop it, but as the default command is only `cat` the container should use less that 10MB of memory.

**Comment by rarkins** [comment]
So.. can you do it? :)

BTW I'm curious if using the `-t` flag for `docker exec` would combine stdout/stderr and make the result more readable for us.
